### ISSUE TITLE
Strange output for BCELoss and BCEWithLogitsLoss

### ISSUE THREAD

**Issue by jyzhang-bjtu** [pytorch/pytorch#4867]
When I use these two loss function or nn modules, they give strange outputs for some extreme values.

Example 1 for BCELoss
input = torch.autograd.Variable(torch.Tensor([1,1]))
target = torch.augograd.Variable(torch.Tensor([0,0]))
loss1 = torch.nn.BCELoss()
a1 = loss1(input,taget)
 The output 
  a1 = 27.6310
 It is very strange. Is the limit of BCELoss is 1e12?

Example 2 for BCEWithLogitsLoss
input = torch.autograd.Variable(torch.Tensor([8e7,8e7]))
target = torch.augograd.Variable(torch.Tensor([0,0]))
loss0 = torch.nn.BCEWithLogitsLoss()
a0 = loss0(input,taget)
 The output 
  a0 = 8e7
Why?

**Comment by bheinzerling** [comment]
BCELoss adds an epsilon of 1e-12 to prevent the undefined log(0)

https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/BCECriterion.c
